Public/Src/Cache/ContentStore/Distributed/NuCache/EventStreaming/EventHubContentLocationEventStore.cs

```diff
@@ -112,30 +112,51 @@ protected override async Task<BoolResult> StartupCoreAsync(OperationContext cont
                 return baseInitializeResult;
             }
 
-            var connectionStringBuilder =
-                new EventHubsConnectionStringBuilder(_configuration.EventHubConnectionString)
-                {
-                    EntityPath = _configuration.EventHubName,
-                };
-
             _currentEventProcessor = new Processor(context, this);
 
-            // Retry behavior in the Azure Event Hubs Client Library is controlled by the RetryPolicy property on the EventHubClient class.
-            // The default policy retries with exponential backoff when Azure Event Hub returns a transient EventHubsException or an OperationCanceledException.
-            _eventHubClient = EventHubClient.CreateFromConnectionString(connectionStringBuilder.ToString());
-            _partitionSender = _eventHubClient.CreatePartitionSender(PartitionId);
+            RegenerateEventHubClientFromCredentials(new EventHubsConnectionStringBuilder(_configuration.EventHubConnectionString)
+            {
+                EntityPath = _configuration.EventHubName,
+            });
 
             return BoolResult.Success;
         }
 
+        private readonly object _eventHubLock = new object();
+
+        public void UpdateCredentials(OperationContext context, EventHubsConnectionStringBuilder connectionStringBuilder)
+        {
+            lock (_eventHubLock)
+            {
+                DoSuspendProcessing(context).ThrowIfFailure();
+                RegenerateEventHubClientFromCredentials(connectionStringBuilder);
+                DoStartProcessing(context, GetLastProcessedSequencePoint()).ThrowIfFailure();

```

Comment:
>DoStartProcessing [](start = 16, length = 17)

this method seems to acquire the log itself already...

Reply:
line 431

---
In reply to: [293123691](https://github.com/microsoft/BuildXL/pull/459#discussion_r293123691) [](ancestors = 293123691)

Reply:
It is done on purpose, since each of these methods acquire the lock separately, and it is possible that this may interleave with a different call to suspend, taking the reentrant lock guarantees that no other call will interleave

---
In reply to: [293123691](https://github.com/microsoft/BuildXL/pull/459#discussion_r293123691) [](ancestors = 293123691)